webapp/packages/core-connections/src/ConnectionInfoOriginResource.ts

```diff
@@ -0,0 +1,114 @@
+import { ProjectsService } from '@cloudbeaver/core-projects';
+import { CachedMapResource, isResourceAlias, ResourceKey, resourceKeyList, ResourceKeyUtils } from '@cloudbeaver/core-resource';
+import { DatabaseConnectionOriginFragment, GraphQLService } from '@cloudbeaver/core-sdk';
+import { schemaValidationError } from '@cloudbeaver/core-utils';
+
+import { CONNECTION_INFO_PARAM_SCHEMA, IConnectionInfoParams } from './CONNECTION_INFO_PARAM_SCHEMA';
+import {
+  ConnectionInfoActiveProjectKey,
+  ConnectionInfoProjectKey,
+  ConnectionInfoResource,
+  createConnectionParam,
+  isConnectionInfoParamEqual,
+} from './ConnectionInfoResource';
+
+@injectable()
+export class ConnectionInfoOriginResource extends CachedMapResource<IConnectionInfoParams, DatabaseConnectionOriginFragment> {
+  constructor(
+    private readonly projectsService: ProjectsService,
+    private readonly graphQLService: GraphQLService,
+    private readonly connectionInfoResource: ConnectionInfoResource,
+  ) {
+    super();
+
+    this.aliases.add(ConnectionInfoProjectKey, param => resourceKeyList(this.keys.filter(key => param.options.projectIds.includes(key.projectId))));
+
+    this.aliases.add(ConnectionInfoActiveProjectKey, () =>
+      resourceKeyList(this.keys.filter(key => projectsService.activeProjects.some(({ id }) => id === key.projectId))),
+    );
+
+    this.sync(this.connectionInfoResource);
+    this.connectionInfoResource.onItemDelete.addHandler(this.delete.bind(this));
+  }
+
+  protected async loader(
+    originalKey: ResourceKey<IConnectionInfoParams>,
+    _: any,
+    refresh?: boolean,
+  ): Promise<Map<IConnectionInfoParams, DatabaseConnectionOriginFragment>> {
+    const connectionsList: DatabaseConnectionOriginFragment[] = [];
+    const projectKey = this.aliases.isAlias(originalKey, ConnectionInfoProjectKey);
+    let removedConnections: IConnectionInfoParams[] = [];
+    let projectId: string | undefined;
+    let projectIds: string[] | undefined;
+
+    if (projectKey) {
+      projectIds = projectKey.options.projectIds;
+    }
+
+    if (this.aliases.isAlias(originalKey, ConnectionInfoActiveProjectKey)) {
+      projectIds = this.projectsService.activeProjects.map(project => project.id);
+    }
+
+    if (isResourceAlias(originalKey)) {
+      const key = this.aliases.transformToKey(originalKey);
+      const outdated = ResourceKeyUtils.filter(key, key => this.isOutdated(key));
+
+      if (!refresh && outdated.length === 1) {
+        originalKey = outdated[0]; // load only single connection
+      }
+    }
+
+    await ResourceKeyUtils.forEachAsync(originalKey, async key => {
+      let connectionId: string | undefined;
+      if (!isResourceAlias(key)) {
+        projectId = key.projectId;
+        connectionId = key.connectionId;
+      }
+
+      const { connections } = await this.graphQLService.sdk.getUserConnectionsOrigin({
+        projectId,
+        connectionId,
+        projectIds,
+      });
```

Comment:
probably we can move such code to helper function and reuse it across resources with same logic:
```
function parseConnectionKey(key) {
  ...
  return { projectId, projectIds, key } // we need to return key because we can modify it sometimes: originalKey = outdated[0]; // load only single connection
```

Reply:
done